shuup/product_discounts/module.py

```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+# This file is part of Shuup.
+#
+# Copyright (c) 2012-2018, Shuup Inc. All rights reserved.
+#
+# This source code is licensed under the OSL-3.0 license found in the
+# LICENSE file in the root directory of this source tree.
+import six
+from django.db.models import Q
+from django.utils.translation import ugettext_lazy as _
+
+from shuup.core.models import Category, ShopProduct
+from shuup.core.pricing import DiscountModule
+
+from .models import ProductDiscount
+
+
+class ProductDiscountModule(DiscountModule):
+    identifier = "product_discounts"
+    name = _("Product Discount Module")
+
+    def discount_price(self, context, product, price_info):
+        shop = context.shop
+        product_id = product if isinstance(product, six.integer_types) else product.pk
+
+        category_ids = Category.objects.filter(shop_products__product_id=product_id).values_list("id", flat=True)
+        group_ids = context.customer.groups.values_list("id", flat=True)
+
+        # Product condition is always applied
+        condition_query = (Q(product__isnull=True) | Q(product_id=product_id))
+
+        # Apply category conditions
+        condition_query &= (
+            Q(category__isnull=True) |
+            (Q(exclude_selected_category=False) & Q(category__id__in=category_ids)) |
+            (Q(exclude_selected_category=True) & ~Q(category__id__in=category_ids))
+        )
+
+        # Apply contact conditions
+        condition_query &= (Q(contact__isnull=True) | Q(contact_id=context.customer.pk))
+
+        # Apply contact group conditions
+        condition_query &= (
+            Q(contact_group__isnull=True) |
+            (Q(exclude_selected_contact_group=False) & Q(contact_group__id__in=group_ids)) |
+            (Q(exclude_selected_contact_group=True) & ~Q(contact_group__id__in=group_ids))
+        )
+
+        # Get all possible discounts for the current product and context
+        potential_discounts = ProductDiscount.objects.available(shop).filter(
+            condition_query).values_list("discounted_price_value", "discount_amount_value", "discount_percentage")
```

Comment:
As we have many (I mean **many**) rows in DB, maybe consider using `iterator` here as, by default, Django loads all results into memory and starts to iterate them through. With `.iterator()` the memory footprint would be way smaller, I guess :)

More info: https://docs.djangoproject.com/en/2.0/ref/models/querysets/#iterator

Reply:
Yes we can use it, but MySQL doesn't support this =)